### ISSUE TITLE
Implement Japanese Azure Images CD Pipeline

### ISSUE THREAD

**Issue by robe070** [robe070/cookbooks#139]
Requires project `P2 Azure` to be completed.
Requires project `P4 AWS Japanese Images` to be completed.

In a similar manner to the Azure English pipeline and the AWS Japanese pipeline , add stages and jobs for Japanese.
This predominantly involves duplicating the English Stages and Jobs and providing different image ids, language identifiers and different test cases.

When duplicating Stages from the English, split stages with multiple jobs into 1 job per stage.

**Comment by robe070** [comment]
This comment is about changing the region tested.

### Preview
There are 18 existing Preview Stages. 6 are Windows2012. These are to be removed soon, so leave them as they are.

The remaining 12 stages are for win2016 and win2019. They test the same 3 regions 4 times. This is to be changed to use a different region for each stage (ignoring win2012 stages).

These are the regions to be used with the text in the order that the stage name needs to be - Country first
Europe West
US North Central
US West
SA North
India Central
Canada Central
France Central
Brazil South
India South
France South
Germany North
UK West

The 4 JPN images will be tested just once. They will REUSE the 4 France Central stages which are currently ENG win2016 and win2019. So new stages will NOT be created for JPN preview tests.

### Production

I have already changed all the production stages to be different regions.

The JPN images will require **1 new stage** for each image/version combination. The regions will be:
US Central
Europe North
India South
Switzerland North

**Comment by ganeshk146** [comment]
Hi Rob,

Please confirm the stages for Preview and Production

Preview Stages
---------------------------
```
Europe West -> w16-14-2
US North Central => w16-15-0
US West -> w16-14-2
SA North -> w16-15-0
India Central -> w16-14-2
Canada Central -> w16-15-0

France Central -> w19-14-2
Brazil South -> w19-15-0
India South -> w19-14-2
France South -> w19-15-0
Germany North -> w19-14-2
UK West -> w19-15-0 


Add entire different regions for testing
w16-14-2j
 w16-15-0j
w19-14-2j
 w19-15-0j
```

Production Stages
---------------------------
US Central -> w16-14-2j
Europe North -> w16-15-0j
India South -> w19-14-2j
Switzerland North -> w19-15-0j

**Comment by ganeshk146** [comment]
In template 2.0 release pipeline existing stages
----------------------------------------------------
AUS SQLAZURE1 DB  -> w12r2d-14-2
AUS SQLAZURE2 DB  -> w19d-14-2
AUS SQLAZURE3 DB  -> w16d-14-2
AUS SQLAZURE4  DB -> w19d-15-0
AUS SQLAZURE5 DB  ->  w16d-15-0
AUS SQLAZURE6 DB  -> w12r2d-15-0
AUS  MYSQL1 and MSSQL DB -> w12r2d-14-2
AUS MSSQLS1 DB  -> w12r2d-14-2

AUS Dev Shoe(With Key Vault Cert) Sizes -> w19d-15-0
AUS Test Shoe(With Key Vault Cert) Sizes -> w19d-15-0
AUS Small Stack Shoe Sizes                        -> w19d-15-0
AUS Medium Stack Shoe Sizes                    -> w19d-15-0
AUS Large Stack Shoe Sizes                        -> w19d-15-0



After Adding JPN Images the list will be 
------------------------------------------------------------------------
AUS SQLAZURE1 DB  -> w12r2d-14-2
AUS SQLAZURE2 DB  -> w19d-14-2
AUS SQLAZURE3 DB  -> w16d-14-2
AUS SQLAZURE4  DB -> w19d-15-0
AUS SQLAZURE5 DB  ->  w16d-15-0
AUS SQLAZURE6 DB  -> w12r2d-15-0
AUS  MYSQL1 and MSSQL DB -> **_w19d-15-0_**
AUS MSSQLS1 DB  -> **_w19d-14-2_**

AUS Dev Shoe(With Key Vault Cert) Sizes -> w19d-15-0
AUS Test Shoe(With Key Vault Cert) Sizes -> w16d-14-2j
AUS Small Stack Shoe Sizes                        -> w19d-15-0j
AUS Medium Stack Shoe Sizes                    -> w16d-15-0j
AUS Large Stack Shoe Sizes                        -> w19d-14-2j

**Comment by robe070** [comment]
I have edited your comment and changed 2 from w12 to w19.

Also note that `AUS MYSQL1 and MSSQL DB` should be called `AUS MYSQL1 DB` and should only contain the MYSQL test

**Comment by Arshpreet1002** [comment]
Hi Rob ,Please close the ticket

**Comment by robe070** [comment]
The pipeline needs to be taken through the publishing process before it can be closed.

**Comment by ganeshk146** [comment]
As per discussion The changes required are 
- Change w2012 stages same as other stages (eg: w2016 and w2019 stages)

**Comment by robe070** [comment]
@ganeshk146 There are more win2012 stages to change later on in that pipeline

**Comment by ganeshk146** [comment]
Hi @robe070 ,

The requested changes are done, Please review it.

**Comment by robe070** [comment]
They look OK. I'll launch a stack with the parameters I need to publish the images

**Comment by robe070** [comment]
Azure Image Release has succeeded up to Preview stage. I had already submitted the images to Azure Marketplace. And Azure Image publication has just reached Preview stage. So I can now continue the pipeline once I've performed the manual tasks. 

Please track the progress of the pipeline and fix any issues that arise.

**Comment by ganeshk146** [comment]
@robe070 ,

Please let me know is there anything left out to close this ticket

**Comment by robe070** [comment]
I'm working on it.

**Comment by robe070** [comment]
The Azure Templates release pipeline has 2 cookbooks artefacts. I think Cookbooks1 is the correct one. Please sort it out and run it to test it. But it needs the alias changed

![image](https://user-images.githubusercontent.com/4089160/130194839-161fd4ab-4562-4912-8b4d-a17aab29c3a0.png)

**Comment by robe070** [comment]
Also, make the same timeout/termination changes on the last task that deletes the Azure Resource Group in Azure pipelines as you have done for AWS.

![image](https://user-images.githubusercontent.com/4089160/130214149-425f8261-fb1f-47f6-9c5d-2890dbcb4a85.png)
![image](https://user-images.githubusercontent.com/4089160/130214202-d2859ac9-a979-4021-bbc8-3f9ebcd21827.png)

**Comment by ganeshk146** [comment]
The job level changes can't be done as the pipeline is not allowing it as a variable, please refer #117 comments.

Other changes i.e Delete stack changes are done 

Template Release Pipeline test https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-environment-logs&releaseId=309&environmentId=5337

Image Release pipeline test 
https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=308

**Comment by robe070** [comment]
As it can't be a variable, it must be a literal. So therefore please assign the value 120 to the job timeout.

Edit: Can you please cancel the releases as this change must be done before testing. You will need to cancel each stage thats queued.

@ganeshk146

**Comment by ganeshk146** [comment]
Template Release 2.0 Pipeline
https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=310


Image Release 2.0 pipeline url 
https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=311

**Comment by robe070** [comment]
Template Release 2.0 Pipeline
I've re-created the release with more tests queued

**Comment by robe070** [comment]
Preview test of w19d-15-0j has failed as the conditions are incorrect. But the gate variable is also wrong. I'm going to address this one failure. @ganeshk146 you need to check all the JPN preview and production tests and ensure the conditions and gate correctly match the SKU embedded in the Stage name e.g. this failing stage `w19d150j Germany West Central Preview` is meant to be testing `w19d-15-0j` but its actually testing `w19d-15-0`

Here is an example of the failure:
https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?releaseId=315&_a=release-environment-logs&environmentId=5530

And I've fixed that one!

Note that the Gate Variable in this example was:
`-Version "w19d-15-0" -osName "Windows Server 2019"`
and needed to be this:
`-Version "w19d-15-0j" -osName "Windows Server 2019 JPN"`

As well as the condition being incorrect. The way to fix the condition on the Task is to REMOVE it altogether. Its on the Job so its superfluous to have it on the Set Gate Variable Task as well.

At the same time, fix up the Task naming. The following image has 2 stages that have the name of the SKU in the title of the task. That is wrong. The SKU must ONLY be in the Stage name. So these other 2 Task names (which in this case are not the same as each other anyway!) must be removed. They should be simply this:

Deploy ARM Template
Delete Resource Group

The stack name is set as a Gate Variable and can change, so it makes no sense to treat it as a constant when its actually a variable.

Do this for EVERY testing stage in Image and Template pipelines.

![image](https://user-images.githubusercontent.com/4089160/130706909-84cc548c-ed27-4f40-afec-1186fd09011b.png)

**Comment by ganeshk146** [comment]
@robe070 ,

The changes are done for Image Release Pipeline 2.0, Working on Template Pipeline 2.0

**Comment by ganeshk146** [comment]
@robe070 ,

The changes are done for Template Pipeline 2.0

**Comment by robe070** [comment]
@ganeshk146 
The Image Release Pipeline is failing for 3 reasons
1. The resource groups were not unique. I have fixed those
2. StackName is not unique. "$(Gate.Sku)"  should be used instead of the current names which have a pattern similar to "yr1614$(Build.BuildNumber)". This needs to be changed for ALL template deployments
3. Japanese Production tests are not running the Japanese tests.

Template Release pipeline may have issues with the stackname. Please extract the stackname from every stage and include it in a post here for review.

Template Release pipeline may have issues with running JPN tests. Please check this.

I will need to run the tests. So please inform me when you have made the changes, particularly the Image Release pipeline which is holding up the template release.

**Comment by robe070** [comment]
Azure image pipeline
Also region UAE Central is not working. It looks like an issue with the tool being used. So change the UAE Central stage to use UAE North. Test this by itself as the only stage being run and ensure that the region works.

**Comment by ganeshk146** [comment]
@robe070 ,
My credentials are not working for Azure Devops login , I request to reset it. my username is "GaneshKumar@lansacloudlansacom.onmicrosoft.com"

**Comment by robe070** [comment]
What is the error message?

**Comment by robe070** [comment]
I have sent the new password in Slack

**Comment by ganeshk146** [comment]
> What is the error message?

Thanks for the new password , Now I am logged in.

**Comment by ganeshk146** [comment]
@robe070 ,

For this issue 
`2. StackName is not unique. "$(Gate.Sku)" should be used instead of the current names which have a pattern similar to "yr1614$(Build.BuildNumber)". This needs to be changed for ALL template deployments
`
It seems like I have forgotten where this need to be changed, please attach a screenshot

**Comment by robe070** [comment]
![image](https://user-images.githubusercontent.com/4089160/134844273-fc7d4d83-683b-4ff0-9706-5100d5a748dc.png)

**Comment by ganeshk146** [comment]
@robe070 ,
The above change needs to be done on this pipeline https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseDefinition?definitionId=17&_a=environments-editor-preview  and for all "AUS", "Preview" and "Production" Stages , Please confirm

**Comment by robe070** [comment]
Correct

**Comment by robe070** [comment]
Another issue just arisen:
w16d150 Aus East Production
2021-09-27 03:15:27Z Verifying the Registry entry for VersionText w16d-15-0-13 and the SKU 
2021-09-27T03:15:46.3824474Z w16d-14-2-13

There was another issue like this I fixed. And there are still 5 or so Production stages that are failing. So please check that these parameters are correct for all the Stages which are failing in PRODUCTION Release-59 (There are failed stages earlier but I have fixed those issues already.). The parameters are:
1. The SKU in the title matches the Gate Variable
2. The Deploy ARM Template parameters LansaVersion and osName are correct and match the SKU.:
-osName "$(Gate.osName)" -lansaVersion "V14 SP2"

**Comment by ganeshk146** [comment]
Sure @robe070

**Comment by robe070** [comment]
Obviously the LansaVersion has two possible values "V14 SP2" and "V15 GA" depending on the SKU value.

**Comment by ganeshk146** [comment]
@robe070 , 

The changes are done, Please run the pipeline.

**Comment by robe070** [comment]
3. JPN image tests is not done.
task `Test ARM Deployment : URL Tests`
requires these arguments for JPN stages:
```
-deploymentOutput '$(deploymentOutput)' -Language $(jpnLanguage)
```

**Comment by ganeshk146** [comment]
The JPN image tests were missing one parameter "-collation $(Collation_JPN)" it is added now

**Comment by ganeshk146** [comment]
> 1. JPN image tests is not done.
>    task `Test ARM Deployment : URL Tests`
>    requires these arguments for JPN stages:
> 
> ```
> -deploymentOutput '$(deploymentOutput)' -Language $(jpnLanguage)
> ```

oh ok, let me check it

**Comment by robe070** [comment]
And there are the Template pipeline to review

> The JPN image tests were missing one parameter "-collation $(Collation_JPN)" it is added now

Thats incorrect
```
-deploymentOutput '$(deploymentOutput)' -Language $(jpnLanguage)
```

**Comment by robe070** [comment]
And the template tasks as well.

I'll be in a meeting 25 mins from now at 3:45 AEST for 1 hour

**Comment by ganeshk146** [comment]
> And there are the Template pipeline to review
> 
> > The JPN image tests were missing one parameter "-collation $(Collation_JPN)" it is added now
> 
> Thats incorrect
> 
> ```
> -deploymentOutput '$(deploymentOutput)' -Language $(jpnLanguage)
> ```

In template release pipeline verified the parameters and it seems correct

**Comment by robe070** [comment]
[Release-61](https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=328) is running the publication stages.

Please keep an eye on it and address any failures.

**Comment by robe070** [comment]
Did you make the UAE Central change?

**Comment by ganeshk146** [comment]
> Azure image pipeline
> Also region UAE Central is not working. It looks like an issue with the tool being used. So change the UAE Central stage to use UAE North. Test this by itself as the only stage being run and ensure that the region works.

This change is done

**Comment by robe070** [comment]
US Central SKU must be wrong - missing JPN?

India South needs a different machine size that exists there. The default does not.

w16d150 Aus East
Template LansaVersion is wrong
Registry entry for VersionText w16d-15-0-13 doesn't match the SKU w16d-14-2-13

**Comment by ganeshk146** [comment]
> US Central SKU must be wrong - missing JPN? `The resource group was missing 'j' at the end`
> 
> India South needs a different machine size that exists there. The default does not.
> 
>
> w16d150 Aus East
> Template LansaVersion is wrong --> `Yeah, just now I fixed it`
> Registry entry for VersionText w16d-15-0-13 doesn't match the SKU w16d-14-2-13

**Comment by ganeshk146** [comment]
India South needs a different machine size that exists there. The default does not.

`Should I need to use different resource group?`

**Comment by robe070** [comment]
I've been researching. India South is not available.

Change to West India. I've checked its available

**Comment by robe070** [comment]
Are you looking at the failures?

**Comment by ganeshk146** [comment]
@robe070 ,

I am not understanding this issue https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-environment-logs&releaseId=331&environmentId=6141 Please help me (I double-check the parameters and the Gate variable values)

**Comment by robe070** [comment]
I've just worked it out. The SKU is being truncated at 9 characters by the template, and that removes the 'j'.

I'm considering a solution

**Comment by robe070** [comment]
I'm going to modify the template to truncate at 10 characters and make sure all the identifiers still work in the template. So nothing more for you to do today on the JPN stages. Just ensure the other stages work correctly that you've been fixing.

**Comment by robe070** [comment]
So I've cancelled all the jobs. And the resource groups will need to all be deleted manually and then re-run all these failing jobs. So not conducive to you working on it too because old names will clash.

So I'll get back to you if there is more for you to work on.

**Comment by robe070** [comment]
@ganeshk146 
You made a change to set `StackName` to `"$(Gate.Sku)"`.

This causes errors in identifiers when deploying the template.

Note that errors only appeared at first when creating 2 stacks where one was Japanese and one was English. On further tweaks it broke with non-unique VM base name in the Scale Set

So I've created a new gate variable `$(Gate.VersionClean)` which needs to be used instead. An example of this is
For Version = w19d-15-0j
VersionClean = w19d150j
That is, the dashes are stripped out and the resultant length is 9 chars or smaller.

So please change all values of `StackName` which are `$(Gate.Sku)` to `$(Gate.VersionClean)`

**Comment by ganeshk146** [comment]
> @ganeshk146
> You made a change to set `StackName` to `"$(Gate.Sku)"`.
> 
> This causes errors in identifiers when deploying the template.
> 
> Note that errors only appeared at first when creating 2 stacks where one was Japanese and one was English. On further tweaks it broke with non-unique VM base name in the Scale Set
> 
> So I've created a new gate variable `$(Gate.VersionClean)` which needs to be used instead. An example of this is
> For Version = w19d-15-0j
> VersionClean = w19d150j
> That is, the dashes are stripped out and the resultant length is 9 chars or smaller.
> 
> So please change all values of `StackName` which are `$(Gate.Sku)` to `$(Gate.VersionClean)`

The requested change is completed. Please run the pipeline

**Comment by ganeshk146** [comment]
@robe070 ,

Please confirm the $(Gate.VersionClean) need to be added in Template Pipeline also? https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseDefinition?definitionId=16&_a=definition-tasks&environmentId=148

**Comment by ganeshk146** [comment]
@robe070 ,

As per ur suggestion, I was updated "UAE North" region 
Please suggest which region we need to use it,. The error says `Provisioning is restricted in this region`
https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-environment-logs&releaseId=337&environmentId=6361
cc: @anant-k-agrawal , @div1127

**Comment by robe070** [comment]
Spot check of VersionClean is all good.

> As per ur suggestion, I was updated "UAE North" region
> Please suggest which region we need to use it,. The error says `Provisioning is restricted in this region`
> https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-environment-logs&releaseId=337&environmentId=6361
> cc: @anant-k-agrawal , @div1127

I cannot determine how to change it. Its getting an error with SQL Server. Yet that provider is available in UAE North.

Try Southeast Asia

**Comment by robe070** [comment]
...and ensure the ResourceGroup variable is appropriately changed

**Comment by robe070** [comment]
Azure Template Release Pipeline needs the stack name change to use Gate.VersionClean. As it stands it will fail because the same name is used on a number of production Stages. Change ALL production tests. 

Looks like AUS Stages are fine as it is. At least they look to be unique. And if you changed to VersionClean it would break because images are used more than once.

**Comment by ganeshk146** [comment]
I am not going to change the Template Release2.0 pipeline as it is running many stages, Once the release finishes execution then I will add the remaining changes
<img width="963" alt="Screenshot 2021-09-28 at 1 28 45 PM" src="https://user-images.githubusercontent.com/72592790/135047241-d73127c6-0122-4083-adb3-ce7a8b2f0686.png">

**Comment by robe070** [comment]
Why does it matter? You can change them at the same time.

**Comment by ganeshk146** [comment]
> Azure Template Release Pipeline needs the stack name change to use Gate.VersionClean. As it stands it will fail because the same name is used on a number of production Stages. Change ALL production tests.
> 
> Looks like AUS Stages are fine as it is. At least they look to be unique. And if you changed to VersionClean it would break because images are used more than once.

This change is completed in Template pipeline 2.0

**Comment by ganeshk146** [comment]
@robe070 ,

All comments are fixed, Is that ok to close this ticket?
cc: @div1127, @anant-k-agrawal

**Comment by robe070** [comment]
all comments are coded. They are not tested.

**Comment by ganeshk146** [comment]
Azure Image release pipeline 2.0 tested -> https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=337

Southeast asia region tested -> https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=340

And there are errors in the Template release 2.0 pipeline and it is not introduced by our changes
https://dev.azure.com/VisualLansa/Lansa%20Azure%20Scalable%20License%20Images/_releaseProgress?_a=release-pipeline-progress&releaseId=343 Please suggest

**Comment by robe070** [comment]
So? Until its all working the issues can't be closed.

**Comment by robe070** [comment]
ps ALL errors since you made changes have been in the pipeline.

**Comment by ganeshk146** [comment]
Thanks, @robe070 , for the support, I learnt a lot from you in terms of coding and also design.
Looking forward to working again with you
